src/coreclr/jit/lowerxarch.cpp

```diff
@@ -1003,6 +1003,154 @@ void Lowering::LowerFusedMultiplyAdd(GenTreeHWIntrinsic* node)
     }
 }
 
+//----------------------------------------------------------------------------------------------
+// TryLowerConstVec: Unfold the Vector to Broadcast->CreateScalarUnsafe->ConstScalar form
+//                   when the constant vector has all the same elements and is an operand
+//                   of an embedded broadcast compatible instruction.
+//
+//  Arguments:
+//     node - The constant vector node
+GenTree* Lowering::TryLowerConstVec(GenTreeVecCon* node)
```

Comment:
It'd be nice if, rather than checking every constant, we only did this check and fix in the case EVEX was supported and the user could itself contain the constant.

Having it integrated as part of the standard containment checks such that the caller could know its a constant and then trivially fix it up to be a broadcast would make this a bit nicer and less impactful to throughput, for example.

Reply:
Thanks for the inputs!

I can remove the `TryCanonizeEmbBroadcastCandicate` method, we might lose some opportunities to enable embedded broadcast when the candidate is at op1. This can a follow-up optimization later, or do we want to handle it in this PR?

and I will move the unfolding logic for constant vector to `IsContainableHWIntrinsicOp`, naively thinking [this part](https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/lowerxarch.cpp#L7382) might be the right place.

Will the above sound  good to you?

Reply:
> I can remove the TryCanonizeEmbBroadcastCandicate method, we might lose some opportunities to enable embedded broadcast when the candidate is at op1. This can a follow-up optimization later, or do we want to handle it in this PR?

We shouldn't lose any support. The `IsContainableHWIntrinsicOp` is already done against `op2` and then `op1` if the intrinsic is commutative.

`IsContainableHWIntrinsicOp` is primarily responsible for doing the checks of "is this containable at all", hence we'd have the validation checks for the `BroadcastScalarToVector###` here: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/lowerxarch.cpp#L7394. The containment checks for vector constants are themselves already handled here: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/lowerxarch.cpp#L7379-L7383

The caller, which in the case of `Add` is https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/lowerxarch.cpp#L7820-L7832, would then need to do any minor fixups to the vector constant if needed.

I expect the simplest thing to do here would be to have some helper `MakeHWIntrinsicSrcContained(node, op2)`. This helper could then do the minor fixup of converting the `CnsVec` into a `BroadcastScalarToVector128(CnsVec)` if `node` supports it and otherwise simply calling `MakeSrcContained`. This helps keep the logic more "pay for play" and isolated to a place where we know its safe to make such transforms.

Reply:
The other thing `IsContainableHWIntrinsicOp` is responsible for is ensuring that the containment is "safe" to do.

That's what the checks for `IsSafeToContainMem` are for here: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/lowerxarch.cpp#L7372 and here: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/lowerxarch.cpp#L7408

And a slightly different and more relaxed check for `IsSafeToMarkRegOptional` here: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/lowerxarch.cpp#L7362

-----------

In lowering, the trees are in evaluation order from the root node to the end following `node->gtNext`. This is different from the actual shape of the tree.

So while we might have a tree:
```
    HWIntrinsic Add
        |-- op1
        \-- op2
```

The actual linear order of IR could be:
```
op1
someOtherNode
op2
someOtherNode
someCallNode
HWIntrinsic Add
```

What containment does is mark that the node will be directly handled by its parent. This primarily means two things:
1. The register allocator will not allocate it a register
2. Codegen will not directly handle the node itself

This has the functional impact of making the node part of its parent and moves the evaluation of it to be part of its parent as well.

For cases like constants, this is always safe, as they can't have a side effect. However, for nodes which might have side effects (such as assigning locals, writing memory, etc), we have to ensure that there isn't any change to the observed order of the side effects. So, we can't "just" contain `op2` even if it was containable as we have to make sure its "safe" to do so as well

Reply:
Yes, I can see your point here, thanks for the detailed elaboration, and the point I was trying to say is actually, we didn't lose the support for embedded broadcast when the candidate is on op1 of a commutive op, we still can, it's just that we will miss the opportunity to use embedded broadcast in this case if op2 can be contained. op2 will be contained first. If op2 cannot be contained, yes, we can contain the broadcast node and enable embedded broadcast. So I referred this case as an 'optimization', this only happens at commutive ops, but I believe it is worth to catch it.

Reply:
Ah, yes.

We don't really have any way, today, to preference one containment over another containment. We could certainly add such support in the future, but I think it'd be better as its own PR

Reply:
Can you also look at the [changes](https://github.com/dotnet/runtime/pull/84821/commits/c916008e6ac7495ce4f22ac0d091d23b0af101b6#diff-8fe1c38ca575e327ad3227c06d009135287a438045cecbe3301d9a633a46cf0cR7403) I made before this conversation, this someway keeps the original ContainCheck logic unchanged.

Reply:
> Can you also look at the [changes](https://github.com/dotnet/runtime/pull/84821/commits/c916008e6ac7495ce4f22ac0d091d23b0af101b6#diff-8fe1c38ca575e327ad3227c06d009135287a438045cecbe3301d9a633a46cf0cR7403) I made before this conversation, this someway keeps the original ContainCheck logic unchanged.

We don't want to make changes to nodes or introduce new nodes within `IsContainableHWIntrinsic`. It's called from various asserts and other places with the general expectation that its non side-effecting.

Reply:
Make sense, I'll update it according to your previous suggestion. Thanks for the inputs!

Reply:
Hi @tannergooding, `MakeHWIntrinsicSrcContained` has been updatesd, can you please review it? Thanks!